pkg/opensearch/client/client.go

```diff
@@ -55,6 +60,28 @@ var newDatasourceHttpClient = func(ds *backend.DataSourceInstanceSettings) (*htt
 	return httpClient, nil
 }
 
+func sigV4Middleware() httpclient.Middleware {
+	return httpclient.NamedMiddlewareFunc("sigv4", func(opts httpclient.Options, next http.RoundTripper) http.RoundTripper {
+		rt, err := sigv4.New(&sigv4.Config{
+			Service:       opts.SigV4.Service,
+			AccessKey:     opts.SigV4.AccessKey,
```

Comment:
ok figured it out I think that this sigv4.New is coming from https://github.com/grafana/grafana-aws-sdk/blob/main/pkg/sigv4/sigv4.go and that we only use the config details we need depending on the auth type: https://github.com/grafana/grafana-aws-sdk/blob/main/pkg/sigv4/sigv4.go#L243

makes sense please ignore me!

Reply:
This is eventually called in the chain using the httpClientOptions created above (called with grafana-plugin-sdk-go's http client provider New right below). 

I may be wrong, but I think this is how this works: 
So `httpClientOptions, err := ds.HTTPClientOptions()` the config options are parsed from the JSON/data source config editor.
Then we add some customizations related to SigV4 right below it including this function you're asking about right here. It just takes something from "opts." But where is "opts" coming from, you ask? 

So when we call `httpClient, err := httpClientProvider.New(httpClientOptions)` here, in `New` we have `clientOpts := p.createClientOptions(opts...)` https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/provider.go#L75
It seems to simply modify a little bit the httpClientOptions we passed in.
Because have passed in httpClientOptions in opensearch-datasource, we enter this case https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/provider.go#L162 and
there we see that it combines the httpClientProvider middleware (anything from the sdk I suppose) with the SigV4 middleware which was already in `clientOpts.Middlewares`. 

Then it takes those mostly-same httpClientOptions and enters `New` here https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/provider.go#L82

Follow `New` to `GetTransport` and you eventually get to the bit where we use the middleware we pass in. https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/http_client.go#L94 Because the mostly-same httpClientOptions were passed in, that's where the `opts` you're asking about comes from.

Sorry this is not a very nice explanation -- I find this very hard to read and follow. Might be able to run this with a debugger to verify this flow.

Reply:
This is eventually called in the chain using the httpClientOptions created above (called with grafana-plugin-sdk-go's http client provider New right below) and httpClientOptions comes from the config. For the env part, some of the code which is reading from the env is from grafana-aws-sdk ( `sigv4.New` call). 

I may be wrong, but I think this is how this works: 
So `httpClientOptions, err := ds.HTTPClientOptions()` the config options are parsed from the JSON/data source config editor.
Then we add some customizations related to SigV4 right below it including this function you're asking about right here. It just takes something from "opts." But where is "opts" coming from, you ask? 

So when we call `httpClient, err := httpClientProvider.New(httpClientOptions)` here, in `New` we have `clientOpts := p.createClientOptions(opts...)` https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/provider.go#L75
It seems to simply modify a little bit the httpClientOptions we passed in.
Because have passed in httpClientOptions in opensearch-datasource, we enter this case https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/provider.go#L162 and
there we see that it combines the httpClientProvider middleware (anything from the sdk I suppose) with the SigV4 middleware which was already in `clientOpts.Middlewares`. 

Then it takes those mostly-same httpClientOptions and enters `New` here https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/provider.go#L82

Follow `New` to `GetTransport` and you eventually get to the bit where we use the middleware we pass in. https://github.com/grafana/grafana-plugin-sdk-go/blob/v0.161.0/backend/httpclient/http_client.go#L94 Because the mostly-same httpClientOptions were passed in, that's where the `opts` you're asking about comes from.

Sorry this is not a very nice explanation -- I find this very hard to read and follow. Might be able to run this with a debugger to verify this flow.

Reply:
I realize I may not have even looked at the right part of your question, let me know if I misunderstood.